File: related_scripts/verification/step_C/verify_C_pcr_calibration.m
%% Step C verification: PCR calibration for January (month=1) only
%  Compares PCA eigenvalues, PC scores, and per-station OLS coefficients
%  with the Python implementation.
%
%  Uses Step B outputs (ESM at stations) and Step A outputs (GHCN obs).
%  Both MATLAB and Python use identical logic:
%    - PCA on NA-windowed ESM grid field for calibration years
%    - No intercept regression on anomalies (Guaita's approach)
%    - Fixed n_pc = 5, 70% train / 30% test split, rng seed = 2026

clear; close all;

%% paths
trace_tas = 'D:\Dataset\DPastCliM-NA\TraCE21k\TraCE-21K-II.monthly.TREFHT.nc';
out_dir = 'D:\Dataset\DPastCliM-NA\verification\step_C\matlab';
if ~exist(out_dir,'dir'), mkdir(out_dir); end

%% parameters
year_cal_min = 1875;
year_cal_max = 1999;
n_pc = 5;
train_frac = 0.7;
rng_seed = 2026;
target_month = 1;  % January

%% load station obs from Step A (Python canonical)
obs = parquetread('D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_obs.parquet');
meta = parquetread('D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_meta.parquet');

%% load ESM grid field (not station-interpolated — raw grid for PCA)
fprintf('Loading TraCE TREFHT for PCA...\n');
trefht = ncread(trace_tas, 'TREFHT');  % (lon, lat, time)
lat = double(ncread(trace_tas, 'lat'));
lon = double(ncread(trace_tas, 'lon'));
time_ka = double(ncread(trace_tas, 'time'));

cal_year_frac = 1950 + time_ka * 1000;
months_since_0 = round(cal_year_frac * 12);
year_arr = int32(floor(months_since_0 / 12));
month_arr = int32(mod(months_since_0, 12) + 1);

tas_map = trefht - 273.15;
clear trefht

% lon 0-360 -> -180..180
lon180 = mod(lon + 180, 360) - 180;
[lon_sorted, lon_idx] = sort(lon180);
tas_map = tas_map(lon_idx, :, :);
[lat_sorted, lat_idx] = sort(lat);
tas_map = tas_map(:, lat_idx, :);

% NA window with padding (matching Python select_na_window defaults)
na_lat = [-180+5, 75+5];  % wide enough
na_lon = [-180-5, -50+5];
lat_mask = lat_sorted >= -90 & lat_sorted <= 90;  % all lat for now
lon_mask = lon_sorted >= na_lon(1) & lon_sorted <= na_lon(2);
esm_lat = lat_sorted(lat_mask);
esm_lon = lon_sorted(lon_mask);
tas_na = tas_map(lon_mask, lat_mask, :);  % (nlon_na, nlat_na, time)

% filter cal period + January
cal_mask = (year_arr >= year_cal_min) & (year_arr <= year_cal_max) & (month_arr == target_month);
fprintf('  Jan cal steps: %d\n', sum(cal_mask));
esm_jan_cal = tas_na(:,:,cal_mask);  % (nlon, nlat, T_cal)
years_jan_cal = year_arr(cal_mask);

nlon = size(esm_jan_cal, 1);
nlat = size(esm_jan_cal, 2);
T_cal = size(esm_jan_cal, 3);

% reshape to (T_cal, nlon*nlat) for PCA — matching Python's (T, ny*nx)
% Python: arr.reshape(T, ny*nx) where arr is (T, ny, nx) -> col order = (lat varies, then lon)
% MATLAB: we have (lon, lat, time) -> reshape to (T, lon*lat) with permute
field_2d = reshape(permute(esm_jan_cal, [3, 2, 1]), T_cal, []);  % (T_cal, nlat*nlon)
% Wait, Python does arr.reshape(T, ny*nx) where arr is (T, lat, lon)
% So the flattening is lat-major (lat index changes first, then lon)
% MATLAB esm_jan_cal is (lon, lat, time) -> permute to (time, lat, lon) then reshape
field_2d = reshape(permute(esm_jan_cal, [3, 2, 1]), T_cal, []);

fprintf('  field_2d: (%d, %d)\n', size(field_2d,1), size(field_2d,2));

%% PCA (SVD-based, matching Python _pca)
mu = mean(field_2d, 1);  % (1, nfeatures)
Xc = field_2d - mu;
[U, S, V] = svd(Xc, 'econ');
singular_vals = diag(S);
scores_full = U * S;  % (T_cal, min(T,F))
var_frac = (singular_vals.^2) / max(sum(singular_vals.^2), 1e-30);

% select top n_pc
eofs = V(:, 1:n_pc)';  % (n_pc, nfeatures) — same as Python's Vt[:n_pc]
pcs = scores_full(:, 1:n_pc);  % (T_cal, n_pc)
ev_top = var_frac(1:n_pc);

fprintf('PCA: top-%d explained variance: [%s]\n', n_pc, num2str(ev_top'*100, '%.2f%% '));

%% save PCA outputs
% eigenvalues
fid = fopen(fullfile(out_dir, 'eigenvalues.bin'), 'w');
fwrite(fid, length(var_frac), 'int32');
fwrite(fid, var_frac, 'double');
fclose(fid);

% PC scores (T_cal x n_pc)
fid = fopen(fullfile(out_dir, 'pc_scores.bin'), 'w');
fwrite(fid, [T_cal, n_pc], 'int32');
fwrite(fid, pcs', 'single');  % transpose for row-major Python reading
fclose(fid);

% field mean
fid = fopen(fullfile(out_dir, 'field_mean.bin'), 'w');
fwrite(fid, length(mu), 'int32');
fwrite(fid, mu, 'single');
fclose(fid);

%% prepare station obs for January regression
fprintf('Preparing station observations...\n');
obs_jan = obs(obs.month == target_month, :);
obs_jan = obs_jan(obs_jan.year >= year_cal_min & obs_jan.year <= year_cal_max, :);

% pivot to (year x station) matrix
[unique_ids, ~, station_idx] = unique(obs_jan.ID);
[unique_years, ~, year_idx] = unique(obs_jan.year);
n_stations_raw = length(unique_ids);
n_years = length(unique_years);

Y_mat = nan(n_years, n_stations_raw, 'single');
for i = 1:height(obs_jan)
    Y_mat(year_idx(i), station_idx(i)) = obs_jan.value(i);
end

% align years with ESM
[common_years, ia, ib] = intersect(unique_years, years_jan_cal);
Y_aligned = Y_mat(ia, :);
pcs_aligned = pcs(ib, :);

% filter stations with >= 30 valid years
valid_count = sum(~isnan(Y_aligned), 1);
keep = valid_count >= 30;
Y_aligned = Y_aligned(:, keep);
station_ids_keep = unique_ids(keep);
n_stations = size(Y_aligned, 2);
fprintf('  stations with >= 30 years: %d / %d\n', n_stations, n_stations_raw);

%% train/test split (deterministic: first 70% years for train)
T_common = size(Y_aligned, 1);
n_train = max(round(T_common * train_frac), 30);
idx_train = 1:n_train;
idx_test = (n_train+1):T_common;
perm = 1:T_common;  % dummy for saving

X_tr = pcs_aligned(idx_train, :);
X_te = pcs_aligned(idx_test, :);
Y_tr = Y_aligned(idx_train, :);
Y_te = Y_aligned(idx_test, :);

%% per-station OLS (WITH intercept — matching our Python implementation)
fprintf('Running per-station OLS (with intercept)...\n');
beta = nan(n_pc+1, n_stations, 'single');
rmse_train = nan(1, n_stations, 'single');
rmse_test = nan(1, n_stations, 'single');
r2_train = nan(1, n_stations, 'single');

X_tr_int = [ones(size(X_tr,1),1), X_tr];  % prepend intercept column
X_te_int = [ones(size(X_te,1),1), X_te];

for s = 1:n_stations
    ys = Y_tr(:, s);
    m = ~isnan(ys);
    if sum(m) < 20
        continue
    end
    % OLS: beta = (X'X)^-1 X'Y
    Xm = X_tr_int(m, :);
    Ym = ys(m);
    b = Xm \ Ym;
    beta(:, s) = single(b);
    
    yhat = Xm * b;
    resid = Ym - yhat;
    rmse_train(s) = sqrt(mean(resid.^2));
    ss_res = sum(resid.^2);
    ss_tot = sum((Ym - mean(Ym)).^2);
    if ss_tot > 0
        r2_train(s) = 1 - ss_res / ss_tot;
    end
    
    % test
    ys_te = Y_te(:, s);
    mt = ~isnan(ys_te);
    if any(mt)
        pred_te = X_te_int(mt, :) * b;
        rmse_test(s) = sqrt(mean((pred_te - ys_te(mt)).^2));
    end
end

fprintf('  median RMSE train: %.4f\n', median(rmse_train, 'omitnan'));
fprintf('  median RMSE test:  %.4f\n', median(rmse_test, 'omitnan'));
fprintf('  median R2 train:   %.4f\n', median(r2_train, 'omitnan'));

%% save regression outputs
fid = fopen(fullfile(out_dir, 'beta.bin'), 'w');
fwrite(fid, [n_pc+1, n_stations], 'int32');
fwrite(fid, beta, 'single');  % already (n_pc+1, n_stations), col-major = station-by-station
fclose(fid);

fid = fopen(fullfile(out_dir, 'rmse_train.bin'), 'w');
fwrite(fid, n_stations, 'int32');
fwrite(fid, rmse_train, 'single');
fclose(fid);

fid = fopen(fullfile(out_dir, 'rmse_test.bin'), 'w');
fwrite(fid, n_stations, 'int32');
fwrite(fid, rmse_test, 'single');
fclose(fid);

% save train/test split indices and station IDs
fid = fopen(fullfile(out_dir, 'split_perm.bin'), 'w');
fwrite(fid, length(perm), 'int32');
fwrite(fid, perm, 'int32');
fclose(fid);

% save station IDs as a text file
fid = fopen(fullfile(out_dir, 'station_ids.txt'), 'w');
for s = 1:n_stations
    fprintf(fid, '%s\n', station_ids_keep{s});
end
fclose(fid);

fprintf('\nStep C outputs saved to: %s\n', out_dir);

File: related_scripts/verification/step_C/verify_C_pcr_calibration.py
"""Step C verification (Python): PCR calibration for January, tas.

Runs identical logic to the MATLAB verify_C script:
  - PCA on NA-windowed ESM grid field for cal years (Jan only)
  - Per-station OLS with intercept
  - Fixed n_pc=5, 70/30 split, rng seed=2026

Then compares eigenvalues, PC scores, regression coefficients, RMSE
with MATLAB outputs from step_C/matlab/.
"""
from __future__ import annotations
import struct
from pathlib import Path

import numpy as np
import pandas as pd
import xarray as xr

from caz.io.trace import load_trace_var, trace_time_to_year_month

OUT_DIR = Path(r"D:\Dataset\DPastCliM-NA\verification\step_C\python")
ML_DIR  = Path(r"D:\Dataset\DPastCliM-NA\verification\step_C\matlab")
TRACE_TAS = Path(r"D:\Dataset\DPastCliM-NA\TraCE21k\TraCE-21K-II.monthly.TREFHT.nc")
OBS_PQ = Path(r"D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_obs.parquet")
META_PQ = Path(r"D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_meta.parquet")

YEAR_MIN, YEAR_MAX = 1875, 1999
N_PC = 5
TRAIN_FRAC = 0.7
RNG_SEED = 2026
TARGET_MONTH = 1


def read_bin_vec(p: Path) -> np.ndarray:
    with open(p, "rb") as f:
        (n,) = struct.unpack("i", f.read(4))
        return np.frombuffer(f.read(), dtype=np.float32)[:n]


def read_bin_dvec(p: Path) -> np.ndarray:
    with open(p, "rb") as f:
        (n,) = struct.unpack("i", f.read(4))
        return np.frombuffer(f.read(), dtype=np.float64)[:n]


def read_bin_mat(p: Path, dtype=np.float32) -> np.ndarray:
    with open(p, "rb") as f:
        r, c = struct.unpack("ii", f.read(8))
        return np.frombuffer(f.read(), dtype=dtype).reshape(r, c)


def main():
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    # ---- load ESM grid field ----
    print("Loading TraCE tas for PCA...")
    da = load_trace_var(TRACE_TAS, "tas")
    lon0 = da["lon"].values
    if lon0.max() > 180:
        new_lon = ((lon0 + 180) % 360) - 180
        da = da.assign_coords(lon=new_lon).sortby("lon")
    if float(da["lat"][0]) > float(da["lat"][-1]):
        da = da.sortby("lat")
    year, month = trace_time_to_year_month(da["time"].values)
    da = da.assign_coords(year=("time", year), month_of_year=("time", month))

    # apply NA window (matching MATLAB: lon [-185,-45], lat all)
    from caz.io.trace import select_na_window
    da = select_na_window(da, lon_min=-180, lon_max=-50, lat_min=-90, lat_max=90, pad=5.0)

    cal_mask = (year >= YEAR_MIN) & (year <= YEAR_MAX) & (month == TARGET_MONTH)
    # recalculate year/month after spatial subset (time dim unchanged)
    da_cal = da.isel(time=cal_mask).compute()
    years_jan_cal = year[cal_mask]
    print(f"  Jan cal steps: {cal_mask.sum()}")

    arr = da_cal.values.astype(np.float32)  # (T, ny, nx)
    T_cal, ny, nx = arr.shape
    # flatten with lat-first order to match MATLAB (column-major: lat varies fastest)
    field_2d = arr.transpose(0, 2, 1).reshape(T_cal, nx * ny)  # (T, nx*ny) lat-first
    print(f"  field_2d: {field_2d.shape}")

    # ---- PCA ----
    mu = field_2d.mean(axis=0)
    Xc = field_2d - mu
    U, s, Vt = np.linalg.svd(Xc, full_matrices=False)
    scores_full = U * s
    var_frac = (s**2) / max(np.sum(s**2), 1e-30)

    eofs = Vt[:N_PC]
    pcs = scores_full[:, :N_PC]
    ev_top = var_frac[:N_PC]
    print(f"PCA: top-{N_PC} EV: {ev_top * 100}")

    # save PCA outputs
    with open(OUT_DIR / "eigenvalues.bin", "wb") as f:
        f.write(struct.pack("i", len(var_frac)))
        f.write(var_frac.astype(np.float64).tobytes())
    with open(OUT_DIR / "pc_scores.bin", "wb") as f:
        f.write(struct.pack("ii", T_cal, N_PC))
        f.write(pcs.astype(np.float32).tobytes())
    with open(OUT_DIR / "field_mean.bin", "wb") as f:
        f.write(struct.pack("i", len(mu)))
        f.write(mu.astype(np.float32).tobytes())

    # ---- station obs ----
    print("Preparing station observations...")
    obs = pd.read_parquet(OBS_PQ)
    obs_jan = obs[(obs["month"] == TARGET_MONTH) &
                  (obs["year"] >= YEAR_MIN) & (obs["year"] <= YEAR_MAX)]
    station_years = obs_jan.pivot_table(index="year", columns="ID", values="value", aggfunc="first")
    common_years = station_years.index.intersection(pd.Index(years_jan_cal))
    station_years = station_years.loc[common_years]

    # align PCs
    pc_idx = pd.Index(years_jan_cal).get_indexer(common_years)
    pcs_aligned = pcs[pc_idx]

    # filter >= 30 valid years
    valid_count = station_years.notna().sum()
    keep = valid_count >= 30
    station_years = station_years.loc[:, keep]
    station_ids = station_years.columns.values
    n_stations = len(station_ids)
    Y = station_years.values.astype(np.float32)
    print(f"  stations >= 30 years: {n_stations}")

    # ---- train/test split (deterministic: first 70% years for train) ----
    T_common = Y.shape[0]
    n_train = max(int(T_common * TRAIN_FRAC), 30)
    idx_train = np.arange(n_train)
    idx_test = np.arange(n_train, T_common)
    perm = np.arange(T_common)

    X_tr = np.column_stack([np.ones(n_train, dtype=np.float32), pcs_aligned[idx_train]])
    X_te = np.column_stack([np.ones(len(idx_test), dtype=np.float32), pcs_aligned[idx_test]])
    Y_tr = Y[idx_train]
    Y_te = Y[idx_test]

    # ---- per-station OLS ----
    print("Running per-station OLS (with intercept)...")
    beta = np.full((N_PC + 1, n_stations), np.nan, dtype=np.float32)
    rmse_train = np.full(n_stations, np.nan, dtype=np.float32)
    rmse_test = np.full(n_stations, np.nan, dtype=np.float32)
    r2_train = np.full(n_stations, np.nan, dtype=np.float32)

    for s in range(n_stations):
        ys = Y_tr[:, s]
        m = np.isfinite(ys)
        if m.sum() < 20:
            continue
        b, _, _, _ = np.linalg.lstsq(X_tr[m], ys[m], rcond=None)
        beta[:, s] = b
        yhat = X_tr[m] @ b
        resid = ys[m] - yhat
        rmse_train[s] = float(np.sqrt(np.mean(resid**2)))
        ss_res = float(np.sum(resid**2))
        ss_tot = float(np.sum((ys[m] - ys[m].mean()) ** 2))
        r2_train[s] = 1 - ss_res / ss_tot if ss_tot > 0 else np.nan

        ys_te = Y_te[:, s]
        mt = np.isfinite(ys_te)
        if mt.any():
            pred_te = X_te[mt] @ b
            rmse_test[s] = float(np.sqrt(np.mean((pred_te - ys_te[mt]) ** 2)))

    print(f"  median RMSE train: {np.nanmedian(rmse_train):.4f}")
    print(f"  median RMSE test:  {np.nanmedian(rmse_test):.4f}")
    print(f"  median R2 train:   {np.nanmedian(r2_train):.4f}")

    # save
    with open(OUT_DIR / "beta.bin", "wb") as f:
        f.write(struct.pack("ii", N_PC + 1, n_stations))
        f.write(beta.tobytes())
    with open(OUT_DIR / "rmse_train.bin", "wb") as f:
        f.write(struct.pack("i", n_stations))
        f.write(rmse_train.tobytes())
    with open(OUT_DIR / "rmse_test.bin", "wb") as f:
        f.write(struct.pack("i", n_stations))
        f.write(rmse_test.tobytes())
    with open(OUT_DIR / "split_perm.bin", "wb") as f:
        f.write(struct.pack("i", len(perm)))
        f.write(perm.astype(np.int32).tobytes())
    with open(OUT_DIR / "station_ids.txt", "w") as f:
        for sid in station_ids:
            f.write(f"{sid}\n")

    print(f"\nPython outputs saved to {OUT_DIR}")

    # ---- compare with MATLAB ----
    print("\n" + "=" * 50)
    print("Comparison: Python vs MATLAB")
    print("=" * 50)

    if not (ML_DIR / "eigenvalues.bin").exists():
        print("MATLAB outputs not found — run verify_C_pcr_calibration.m first")
        return

    # eigenvalues
    ml_ev = read_bin_dvec(ML_DIR / "eigenvalues.bin")
    py_ev = var_frac
    n_comp = min(len(ml_ev), len(py_ev), 20)
    print(f"\nEigenvalues (top {n_comp}):")
    print(f"  max |diff|:  {np.max(np.abs(py_ev[:n_comp] - ml_ev[:n_comp])):.2e}")
    print(f"  mean |diff|: {np.mean(np.abs(py_ev[:n_comp] - ml_ev[:n_comp])):.2e}")

    # PC scores (note: sign may be flipped)
    ml_pcs = read_bin_mat(ML_DIR / "pc_scores.bin")
    py_pcs = pcs.astype(np.float32)
    print(f"\nPC scores shape: Python {py_pcs.shape}, MATLAB {ml_pcs.shape}")
    if py_pcs.shape == ml_pcs.shape:
        for k in range(N_PC):
            corr = np.corrcoef(py_pcs[:, k], ml_pcs[:, k])[0, 1]
            diff = np.abs(py_pcs[:, k] - ml_pcs[:, k])
            # check if sign-flipped
            corr_flip = np.corrcoef(py_pcs[:, k], -ml_pcs[:, k])[0, 1]
            sign_note = " (sign-flipped)" if abs(corr_flip) > abs(corr) else ""
            print(f"  PC{k+1}: corr={max(abs(corr),abs(corr_flip)):.6f}, "
                  f"max|diff|={diff.max():.4f}{sign_note}")

    # field mean
    ml_mu = read_bin_vec(ML_DIR / "field_mean.bin")
    py_mu = mu.astype(np.float32)
    if len(ml_mu) == len(py_mu):
        dm = np.abs(py_mu - ml_mu)
        print(f"\nField mean: max|diff|={dm.max():.6f}, mean|diff|={dm.mean():.6f}")

    # station IDs overlap
    ml_ids = [l.strip() for l in open(ML_DIR / "station_ids.txt").readlines()]
    py_ids = list(station_ids)
    common_ids = set(py_ids) & set(ml_ids)
    print(f"\nStations: Python {len(py_ids)}, MATLAB {len(ml_ids)}, "
          f"overlap {len(common_ids)}")

    # RMSE comparison for common stations
    if common_ids:
        ml_rmse_tr = read_bin_vec(ML_DIR / "rmse_train.bin")
        py_order = [py_ids.index(sid) for sid in ml_ids if sid in common_ids]
        ml_order = [ml_ids.index(sid) for sid in ml_ids if sid in common_ids]
        py_rmse_sub = rmse_train[py_order[:len(ml_order)]]
        ml_rmse_sub = ml_rmse_tr[ml_order[:len(py_order)]]
        d = np.abs(py_rmse_sub - ml_rmse_sub)
        valid = np.isfinite(d)
        if valid.any():
            print(f"\nRMSE train (common stations):")
            print(f"  max|diff|:  {d[valid].max():.6f}")
            print(f"  mean|diff|: {d[valid].mean():.6f}")


if __name__ == "__main__":
    main()
